libraries/AP_Proximity/AP_Proximity_MAV.cpp

```diff
@@ -166,10 +169,83 @@ void AP_Proximity_MAV::handle_msg(const mavlink_message_t &msg)
 
         // update proximity sectors validity and boundary point
         for (uint8_t i = 0; i < PROXIMITY_NUM_SECTORS; i++) {
-            _distance_valid[i] = (_distance[i] >= _distance_min) && (_distance[i] <= _distance_max);
+            boundary.mark_distance_valid((boundary.get_distance(i) >= _distance_min) && (boundary.get_distance(i) <= _distance_max), i);
             if (sector_updated[i]) {
-                update_boundary_for_sector(i, false);
+                boundary.update_boundary(i);
             }
         }
     }
+
+    if (msg.msgid == MAVLINK_MSG_ID_OBSTACLE_DISTANCE_3D) {
+        mavlink_obstacle_distance_3d_t packet;
+        mavlink_msg_obstacle_distance_3d_decode(&msg, &packet);
+
+        const uint32_t previous_sys_time = _last_update_ms;
+        _last_update_ms = AP_HAL::millis();
+        // time_diff will check if the new message arrived significantly later than the last message
+        const uint32_t time_diff = _last_update_ms - previous_sys_time;
+
+        const uint32_t previous_msg_timestamp = _last_3d_msg_update_ms;
+        _last_3d_msg_update_ms = packet.time_boot_ms;
+        bool clear_fence = false;
+        
+        // we will add on to the last fence if the time stamp is the same
+        // provided we got the new obstacle in less than PROXIMITY_3D_MSG_TIMEOUT_MS  
+        if ((previous_msg_timestamp != _last_3d_msg_update_ms) || (time_diff > PROXIMITY_3D_MSG_TIMEOUT_MS)) {
+            clear_fence = true;
+        }
+
+        _distance_min = packet.min_distance;
+        _distance_max = packet.max_distance;
+        
+        Vector3f current_pos;
+        Matrix3f body_to_ned;
+        const bool database_ready = database_prepare_for_push(current_pos, body_to_ned);
+        const float MAX_DISTANCE = 9999.0f;
+
+        if (clear_fence) {
+            // cleared fence back to defaults since we have a new timestamp
+            for (uint8_t i=0; i < PROXIMITY_NUM_STACK; i++) {
+                for (uint8_t j=0; j < PROXIMITY_NUM_SECTORS; j++) {
+                    boundary.set_angle(boundary._sector_middle_deg[j], j, i);
+                    boundary.set_pitch(boundary._pitch_middle_deg[i], j, i);
+                    boundary.set_distance(MAX_DISTANCE, j, i);
+                    boundary.mark_distance_valid(false, j, i);
+                }
+            } 
+        }
+        
+        const float x = packet.x;
```

Comment:
doesn't really matter but it's probably not necessary to have the x,y,z local variables